doc/installationguide.md

```diff
@@ -21,21 +21,51 @@ These are the parameters that can be configured in the global section:
+* **stats**: configure the periodic collection of statistics. Use `interval` in milliseconds to set the time between stats writings.
 ```
     stats: {
         interval: 100
     }
 ```
-* **authentication**: authentication data, for use in retrieving tokens for devices with a trust token (just needed in scenarios with security enabled in the Context Broker side). E.g.:
-```
-	{
-        host: 'localhost',
-        port: '5000',
-        user: 'iotagent',
-        password: 'iotagent'
-	}
-```
+* **authentication**: authentication data, for use in retrieving tokens for devices with a trust token (required in scenarios with security enabled in the Context Broker side). 
+  Currently, two authentication provider are supported: `keystone` and `oauth2`. 
+  In `keystone` based authentication, the `trust` associated to the `device` or `deviceGroup` is a token representing a specific user and his rights on a given domain (i.e.
+  combination of `fiware-service` and `service-path`). The authentication process use the trust delegation workflow to check if the trust provided is valid, in which
+  case return a `x-subject-token` that can be used to authenticate the request to the context broker. Required parameters are: the `url` of the keystone to be used
+  (alternatively `host` and `port` but if you use this combination, the IoT Agent will assume that the protocol is HTTP), the `user` and `password` to which it is delegated
+  the `trust` verification.
+  E.g.:
+    ```
+    {
+          enabled: true,
+          url: 'https://localhost:5000',
+          type: 'keystone',
+          user: 'iotagent',
+          password: 'iotagent'
+    }
+    ```
+  In `oauth2` based authentication, the `trust` associated to the `device` or `deviceGroup` is a `refresh_token` issued by a specific user for the context broker client. 
+  The authentication process use the [`refresh_token` grant type](https://tools.ietf.org/html/rfc6749#section-1.5) to obtain an `access_token`
+  that can be used to authenticate the request to the context broker.
+  At the time being the assumption is that the `refresh_token` is a not expiring `offline_token` (we believe this is the best solution in the case of IoT Devices,
+  since injecting a refresh token look may slow down communication. Still the developer would be able to invalidate the refresh token on the provider side in
+  case of security issues connected to a token). The code was tested using Keycloak and Auth0 (it may require customisation
+  for other providers - while OAuth2 is a standard, not all implementations behave in the same way, especially as regards status codes and error messages).
+  Required parameters are: the `url` of the OAuth 2 provider to be used (alternatively `host` and `port` but if you use this combination, the IoT Agent will assume
+  that the protocol is HTTP), the `tokenPath` to which the validation request should be sent, the `clientId` and `clientSecret` that identify the context broker,
```

Comment:
Reading

> the `clientId` and `clientSecret` that identify the context broker

it seems that Context Broker acts as client in the OAuth athentication. However, I'd say that IOTA takes that role (because is the one that sends the authentication request). Should the documentation be improved? Or maybe I'm missunderstanding how it works (I'm not an expert in OAuth)

Reply:
hi @fgalan OAuth authentication supports different type of flows that allows for different type of authentication.
in this case, being the purpose to have a token specific for the service (each service / application supported by an OAuth provider can be issued specific tokens), we use the [client credentials flow](https://tools.ietf.org/html/rfc6749#section-1.3.4). In this picture, the client (IOTA) authenticate to the OAuth provider using a `clientId` and `clientSecret` that are specific to the service to be accessed (i.e. the context broker). Naming are the conventional names adopted in OAuth.

Reply:
I'm not an expert in OAuth ;) but if people familiar with find that nomenclature is ok, then fine with it.

NTC